### ISSUE TITLE
ORPO training produces a structurally-zero gradient: get_logps() is called outside the value_and_grad-differentiated closure

### ISSUE THREAD

**Issue by jecochard** [Goekdeniz-Guelmez/mlx-lm-lora#68]
**Environment**
- mlx-lm-lora: 3.0.0
- mlx: 0.32.0 / mlx-lm: 0.31.3
- macOS 26.5.2, Qwen/Qwen3-8B, LoRA rank 16

**Root cause**

In `mlx_lm_lora/trainer/orpo_trainer.py`, inside `train_orpo()`'s `step()`:

    chosen_logps, chosen_logits_mean = get_logps(model, chosen, chosen_masks)
    rejected_logps, rejected_logits_mean = get_logps(model, rejected, rejected_masks)

    (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
        chosen_logps, chosen_logits_mean, rejected_logps, rejected_logits_mean,
        chosen_masks, rejected_masks, preference_scores=preference_scores,
    )

`get_logps(model, ...)` -- the only place `model` is actually called
forward -- happens BEFORE and OUTSIDE `loss_wrapper`, which is the
function `nn.value_and_grad(model, loss_wrapper)` differentiates.
`loss_wrapper` itself only receives the already-computed numeric log-
probabilities as plain arguments; it never calls `model`. Since the
differentiated function has no computational dependency on `model`'s
parameters, the gradient computed is exactly zero for every parameter,
every step -- correct given what's actually being differentiated, just
not what's intended.

Contrast with `dpo_trainer.py`'s `train_dpo()`, where `get_token_scores
(model, ...)` is called INSIDE `loss_wrapper`:

    def loss_wrapper(chosen, rejected, chosen_masks, rejected_masks):
        policy_chosen_scores = get_token_scores(model, chosen, chosen_masks)
        policy_rejected_scores = get_token_scores(model, rejected, rejected_masks)
        ...
        return loss_fn(...)

    loss_value_and_grad = nn.value_and_grad(model, loss_wrapper)

Here the forward pass is inside the traced closure, so gradients
correctly flow back to the model. This is why DPO trains correctly under
otherwise identical conditions while ORPO does not.

**Observed symptom**

Training with `--train-mode orpo`, resumed from a converged SFT adapter
via `--resume-adapter-file`, ran to completion over 1,000 iterations
reporting plausible, varying metrics throughout (loss, chosen_r,
rejected_r, margin, accuracy). But the saved adapter -- checked via
SHA-256 at multiple checkpoints including iterations 100, 500, and 1000
-- was byte-for-byte identical to the resumed starting adapter every
time. No parameter update was ever persisted, consistent with a
structurally-zero gradient.

**Fix (verified working)**

Move the `get_logps(model, ...)` calls inside `loss_wrapper`, mirroring
how `dpo_trainer.py` calls `get_token_scores` inside its own
`loss_wrapper`:

    def loss_wrapper(
        chosen, rejected, chosen_masks, rejected_masks, preference_scores,
    ):
        chosen_logps, chosen_logits_mean = get_logps(model, chosen, chosen_masks)
        rejected_logps, rejected_logits_mean = get_logps(
            model, rejected, rejected_masks
        )
        return loss(
            chosen_logps=chosen_logps,
            chosen_logits_mean=chosen_logits_mean,
            rejected_logps=rejected_logps,
            rejected_logits_mean=rejected_logits_mean,
            chosen_masks=chosen_masks,
            rejected_masks=rejected_masks,
            preference_scores=preference_scores,
            beta=args.beta,
        )

    loss_value_and_grad = nn.value_and_grad(model, loss_wrapper)

    @partial(mx.compile, inputs=state, outputs=state)
    def step(batch, prev_grad, do_update):
        chosen, rejected, chosen_masks, rejected_masks, preference_scores = batch
        (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
            chosen, rejected, chosen_masks, rejected_masks,
            preference_scores=preference_scores,
        )

Applied this patch locally and re-ran the exact reproduction above
(resume from the same SFT adapter, `--train-mode orpo`, 10-20
iterations). Before the patch: saved adapter byte-identical to the
resumed starting checkpoint at every iteration tested. After the patch:
the checkpoint at iteration 10 already differs (confirmed via SHA-256),
with accuracy climbing from 0.5 -> 1.0 and margin moving from ~0.01 to
~0.10-0.15 over the first 20 iterations -- real, working preference
learning.

**Reproduction**

    mlx_lm_lora.train -c orpo_config.yaml --train --train-mode orpo \
      --data ./build/orpo --model Qwen/Qwen3-8B --adapter-path ./adapters/v4 \
      --resume-adapter-file ./adapters/v4-sft/adapters.safetensors

    shasum -a 256 ./adapters/v4/adapters.safetensors ./adapters/v4-sft/adapters.safetensors
    # -> identical at every checkpoint before the fix; differ after it

Happy to open a PR with this exact change if useful.

**Comment by Goekdeniz-Guelmez** [comment]
closed since its the same issue as #65 as well
